File: stats.js
/**
 * Statistics for Produco
 * Calculate productivity metrics
 */

class Statistics {
  constructor(tasks, sessions) {
    this.tasks = tasks;
    this.sessions = sessions;
  }

  getCompletionRate() {
    if (this.tasks.length === 0) return 0;
    const completed = this.tasks.filter(t => t.completed).length;
    return (completed / this.tasks.length) * 100;
  }

  getTotalFocusTime() {
    return this.sessions.reduce((total, session) => total + session.durationMinutes, 0);
  }

  getAverageSessionLength() {
    if (this.sessions.length === 0) return 0;
    return this.getTotalFocusTime() / this.sessions.length;
  }

  getStreak() {
    if (this.sessions.length === 0) return 0;
    
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    
    let streak = 0;
    let currentDate = today;
    
    const sessionDates = this.sessions
      .map(s => new Date(s.endedAt).toDateString())
      .filter((date, index, self) => self.indexOf(date) === index)
      .sort((a, b) => new Date(b) - new Date(a));
    
    for (const dateStr of sessionDates) {
      const sessionDate = new Date(dateStr);
      sessionDate.setHours(0, 0, 0, 0);
      
      const diffDays = Math.floor((currentDate - sessionDate) / (1000 * 60 * 60 * 24));
      
      if (diffDays <= 1) {
        streak++;
        currentDate = sessionDate;
      } else {
        break;
      }
    }
    
    return streak;
  }

  getTasksByPriority() {
    const priorities = { high: 0, medium: 0, low: 0 };
    this.tasks.forEach(task => {
      if (priorities[task.priority] !== undefined) {
        priorities[task.priority]++;
      }
    });
    return priorities;
  }

  getWeeklyFocusTime() {
    const now = new Date();
    const weekAgo = new Date(now.getTime() - 7 * 24 * 60 * 60 * 1000);
    
    return this.sessions
      .filter(s => new Date(s.endedAt) >= weekAgo)
      .reduce((total, session) => total + session.durationMinutes, 0);
  }

  getSummary() {
    return {
      totalTasks: this.tasks.length,
      completedTasks: this.tasks.filter(t => t.completed).length,
      completionRate: this.getCompletionRate(),
      totalFocusMinutes: this.getTotalFocusTime(),
      averageSessionMinutes: this.getAverageSessionLength(),
      streak: this.getStreak(),
      weeklyFocusMinutes: this.getWeeklyFocusTime(),
      tasksByPriority: this.getTasksByPriority()
    };
  }
}

module.exports = Statistics;

File: stats.cpp
/**
 * Statistics calculation for Produco - C++ implementation
 * Performance-optimized statistics
 */

#include <iostream>
#include <vector>
#include <string>
#include <cmath>
#include <algorithm>

struct Task {
    int id;
    bool completed;
};

struct Session {
    int durationMinutes;
    std::string endedAt;
};

class Statistics {
private:
    std::vector<Task> tasks;
    std::vector<Session> sessions;
    
public:
    Statistics(const std::vector<Task>& t, const std::vector<Session>& s) 
        : tasks(t), sessions(s) {}
    
    double getCompletionRate() {
        if (tasks.empty()) return 0.0;
        
        int completed = 0;
        for (const auto& task : tasks) {
            if (task.completed) completed++;
        }
        
        return (static_cast<double>(completed) / tasks.size()) * 100.0;
    }
    
    int getTotalFocusTime() {
        int total = 0;
        for (const auto& session : sessions) {
            total += session.durationMinutes;
        }
        return total;
    }
    
    double getAverageSessionLength() {
        if (sessions.empty()) return 0.0;
        
        return static_cast<double>(getTotalFocusTime()) / sessions.size();
    }
    
    int getStreak() {
        if (sessions.empty()) return 0;
        
        int streak = 0;
        // Simplified streak calculation
        // In production, this would parse dates and calculate consecutive days
        for (size_t i = 0; i < sessions.size(); i++) {
            streak++;
        }
        
        return streak;
    }
    
    void printSummary() {
        std::cout << "=== Statistics Summary ===" << std::endl;
        std::cout << "Total tasks: " << tasks.size() << std::endl;
        std::cout << "Completion rate: " << getCompletionRate() << "%" << std::endl;
        std::cout << "Total focus time: " << getTotalFocusTime() << " minutes" << std::endl;
        std::cout << "Average session: " << getAverageSessionLength() << " minutes" << std::endl;
        std::cout << "Streak: " << getStreak() << " days" << std::endl;
    }
};

int main() {
    std::vector<Task> tasks = {
        {1, true},
        {2, false},
        {3, true}
    };
    
    std::vector<Session> sessions = {
        {25, "2024-01-01"},
        {30, "2024-01-02"}
    };
    
    Statistics stats(tasks, sessions);
    stats.printSummary();
    
    return 0;
}
